### PR TITLE
DPMMA-2550 Point Groups API

### PR DESCRIPTION
**Pull request by patrykgruszka**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ Y ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ Y ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation-new/pull/185
| Issue(s) addressed                     | Fixes #... 

#### Description:
This PR adds API endpoints for Point Groups management.
```
mautic_api_pointGroups_getall          GET     /api/points/groups                                                                               
mautic_api_pointGroups_getone          GET     /api/points/groups/{id}                                                                          
mautic_api_pointGroups_new             POST    /api/points/groups/new                                                                                                                                        
mautic_api_pointGroups_editput         PUT     /api/points/groups/{id}/edit                                                                     
mautic_api_pointGroups_editpatch       PATCH   /api/points/groups/{id}/edit                                                                                                                           
mautic_api_pointGroups_delete          DELETE  /api/points/groups/{id}/delete                                                                   
mautic_api_getcontactpointgroups       GET     /api/contacts/{contactId}/points/groups                                                          
mautic_api_getcontactpointgroup        GET     /api/contacts/{contactId}/points/groups/{groupId}                                                
mautic_api_adjustcontactgrouppoints    POST    /api/contacts/{contactId}/points/groups/{groupId}/{operator}/{value} 
```

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))

2. **Create a new point group**:
   - **Endpoint**: `POST https://your-api-host.com/api/points/groups/new`
   - **Payload**: JSON payload with the following fields:
     ```json
     {
         "name": "New Point Group",
         "description": "Description of the new point group"
     }
     ```

3. **Retrieve all point groups**:
   - **Endpoint**: `GET https://your-api-host.com/api/points/groups`

4. **Update the created point group**:
   - **Endpoint**: `PATCH https://your-api-host.com/api/points/groups/{id}/edit`
   - **Payload**: JSON payload with the fields to update:
     ```json
     {
         "name": "Updated Point Group Name",
         "description": "Updated description of the point group"
     }
     ```

5. **Delete the created point group**:
   - **Endpoint**: `DELETE https://your-api-host.com/api/points/groups/{id}/delete`

6. **Get contact group points**:
   - **Endpoint**: `GET https://your-api-host.com/api/contacts/{leadId}/points/groups`

7. **Adjust contact group points**:
   - **Plus operation**:
     - **Endpoint**: `POST https://your-api-host.com/api/contacts/{leadId}/points/groups/{groupId}/plus/{value}`
   - **Minus operation**:
     - **Endpoint**: `POST https://your-api-host.com/api/contacts/{leadId}/points/groups/{groupId}/minus/{value}`
   - **Set operation**:
     - **Endpoint**: `POST https://your-api-host.com/api/contacts/{leadId}/points/groups/{groupId}/set/{value}`
   - **Divide operation**:
     - **Endpoint**: `POST https://your-api-host.com/api/contacts/{leadId}/points/groups/{groupId}/divide/{value}`
   - **Multiply operation**:
     - **Endpoint**: `POST https://your-api-host.com/api/contacts/{leadId}/points/groups/{groupId}/times/{value}`

When testing these endpoints, ensure that you replace placeholders such as `{leadId}`, `{groupId}`, and `{value}` with actual IDs and values as required for your specific test cases. Additionally, for endpoints that require a request body, provide the appropriate JSON payload with the necessary fields and values.

### CODE REVIEW

**Review by RCheesley — DISMISSED** [review]
Tested all endpoints (thank you SO MUCH for the awesome documentation in the PR @patrykgruszka it really helps testers when you provide both the path and the snippet!).

All working as expected. Thanks for the great work! 🚀

File: app/bundles/PointBundle/Config/config.php [hunk 1]
```diff
@@ -63,6 +63,25 @@
                 'controller' => 'Mautic\PointBundle\Controller\Api\PointApiController::adjustPointsAction',
                 'method'     => 'POST',
             ],
+            'mautic_api_pointgroupsstandard' => [
+                'standard_entity' => true,
+                'name'            => 'pointGroups',
+                'path'            => '/points/groups',
+                'controller'      => \Mautic\PointBundle\Controller\Api\PointGroupsApiController::class,
```
Comment (escopecz) [inline app/bundles/PointBundle/Config/config.php]: As per the new CS Fixer version:
```suggestion
                'controller'      => Mautic\PointBundle\Controller\Api\PointGroupsApiController::class,
```

File: app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+ * @extends CommonApiController<Group>
+ */
+class PointGroupsApiController extends CommonApiController
+{
+    /**
+     * @var PointGroupModel
+     */
+    protected $model;
+
+    /** @phpstan-ignore-next-line */
```
Comment (escopecz) [inline app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php]: Why is this needed? If it cannot be resolved, could you add the reason to the comment?

File: app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+/**
+ * @extends CommonApiController<Group>
+ */
+class PointGroupsApiController extends CommonApiController
+{
+    /**
+     * @var PointGroupModel
+     */
+    protected $model;
```
Comment (escopecz) [inline app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php:36]:
```suggestion
    private PointGroupModel $model;
```
Or is there some weird inheritance thing going on?

File: app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php [hunk 4]
```diff
@@ -0,0 +1,150 @@
+<?php
+
```
Comment (escopecz) [inline app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/PointBundle/Tests/Controller/Api/PointGroupsApiControllerTest.php [hunk 5]
```diff
@@ -0,0 +1,182 @@
+<?php
+
```
Comment (escopecz) [inline app/bundles/PointBundle/Tests/Controller/Api/PointGroupsApiControllerTest.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/PointBundle/Tests/Controller/Api/PointGroupsApiControllerTest.php [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        $this->assertArrayHasKey('pointGroup', $responseData);
+        $updatedData = $responseData['pointGroup'];
+        $this->assertEquals('Updated Point Group Name', $updatedData['name']);
+        $this->assertEquals('Updated description of the point group', $updatedData['description']);
+
+        // Delete the created point group
+        $this->client->request('DELETE', "/api/points/groups/{$createdData['id']}/delete");
+        $deleteResponse = $this->client->getResponse();
+
+        $this->assertSame(200, $deleteResponse->getStatusCode());
```
Comment (escopecz) [inline app/bundles/PointBundle/Tests/Controller/Api/PointGroupsApiControllerTest.php]:
```suggestion
        $this->assertSame(Response::HTTP_OK, $deleteResponse->getStatusCode());
```

**Review by escopecz — CHANGES_REQUESTED** [review]
The code looks great! I have a few small suggestions. The functional tests are testing the happy paths. It would be great to have tests that check also 404s, xss and similar. But happy path tests are still way better than none 👍

File: app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php (hunk 3 above)
Comment (patrykgruszka) [reply app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php:36]: Yes, we can't define the type here.

File: app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php (hunk 2 above)
Comment (patrykgruszka) [reply app/bundles/PointBundle/Controller/Api/PointGroupsApiController.php]: Sure, I've added a comment. It is caused by the parent class.

**Comment by patrykgruszka** [comment]
Thanks! I've updated the code and added some extra tests :)

Would you have some time to look at the related PR in the API library? https://github.com/mautic/api-library/pull/322

**Review by escopecz — APPROVED** [review]
I see no issue in the code. The Point group/score is covered with functional tests 👍

**Pull request merged**